### DISCUSSION TITLE
Collect remaining fields for `biome_deserialize`

### DISCUSSION THREAD

**Discussion by NicholasLYang** [biomejs/biome#2698] [Ideas]
Hi,

I'm looking to use biome_deserialize_macros to parse a `package.json` file. We want to collect any extra fields for other tools in a map, so if we serialize the file back to text, we preserve those extra fields. Is that possible with the current macro? Basically I would like serde's `#[flatten]` but for biome.

**Comment by Conaclos** [comment]
It is currently not possible. How `#[flatten]` solves the issue? `#[flatten]` is to flatten nested structures? Have you a code example?

**Reply by NicholasLYang** [reply]
In this case the nested structure would be a `HashMap`. This is what serde does: https://github.com/serde-rs/serde/issues/941

**Reply by NicholasLYang** [reply]
https://serde.rs/attr-flatten.html#capture-additional-fields

**Reply by NicholasLYang** [reply]
@Conaclos, do you think this is feasible? It's currently blocking our ability to adopt biome for parsing package.json files.

**Reply by Conaclos** [reply]
I think the current structure doesn't allow implementing easily `flatten` as serde supports it. However, we could implement a subset of the behavior of `flatten` by requiring that the type of the annotated field implements `FromIterator<(K, V)` (this is implemented by HashMap). What do you think? Would you like to contribute on this?

**Reply by Conaclos** [reply]
Side note: I am not sure what could be the type of the value if we didn't deserialize them...

**Reply by NicholasLYang** [reply]
Yeah that sounds good! I'll look into implementing it. Could the type of the value be `AnyJsonValue`?

Edit: Hmm probably not since we need it to be format agnostic. Perhaps we could create a biome equivalent of `serde::Value`?

**Reply by NicholasLYang** [reply]
First pass at PR here: https://github.com/biomejs/biome/pull/2757

**Comment by Conaclos** [comment]
This was implemented.
